### PR TITLE
feat: class multi select (RDFA-546)

### PR DESCRIPTION
**Pull request by rema-soptim**
## Checklist

- [ ] Tests added or updated (or not applicable)
- [ ] Documentation updated (or not applicable)
- [ ] No breaking changes introduced (or described in the summary above)
- [ ] Commits are signed off (`git commit -s`) for DCO

### CODE REVIEW

File: backend/src/main/java/org/rdfarchitect/services/dl/update/classlayout/CustomDiagramLayoutUseCase.java [hunk 1]
```diff
@@ -31,8 +31,21 @@ void addClassesToCustomDiagram(
     void removeClassFromCustomDiagram(
             GraphIdentifier graphIdentifier, UUID diagramUUID, UUID classUUID);
 
+    default void removeClassesFromCustomDiagram(
+            GraphIdentifier graphIdentifier, UUID diagramUUID, List<UUID> classUUIDs) {
+        classUUIDs.forEach(
+                classUUID -> removeClassFromCustomDiagram(graphIdentifier, diagramUUID, classUUID));
+    }
+
```
Comment (spah-soptim) [inline backend/src/main/java/org/rdfarchitect/services/dl/update/classlayout/CustomDiagramLayoutUseCase.java]: This calls removeClassFromCustomDiagram once per uuid, opening a separate WRITE transaction per class, so removing N classes does N commits.
Maybe move the loop into the service so the whole batch runs in one transaction.

File: frontend/src/routes/delete-relations-dialog/DeleteDependenciesDialog.svelte [hunk 2]
```diff
@@ -143,11 +164,12 @@
     }
 
     async function submitDeleteRequest() {
```
Comment (spah-soptim) [inline frontend/src/routes/delete-relations-dialog/DeleteDependenciesDialog.svelte:177]: The delete payload is built per affected-resource tree and concatenated with no UUID de-duplication.
When multi-selection contains resources whose dependency trees overlap (for example parent class + subclass), the same uuid is submitted multiple times with potentially conflicting actions (chosen KEEP silently overridden by a DELETE from another tree).

File: frontend/src/lib/rendering/svelteflow/layout/elkLayout.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    "elk.spacing.edgeEdge": "15",
+    "elk.layered.spacing.nodeNodeBetweenLayers": "80",
+    "elk.spacing.nodeNode": "60",
+};
+
+export async function getLayoutedNodes(nodes, edges) {
+    const elk = new ELK({
+        //WEB WORKER: layouting is executed in a separate thread, preventing the frontend from blocking
+        workerFactory: () => new Worker(ElkWorkerURL, { type: "classic" }),
+    });
```
Comment (spah-soptim) [inline frontend/src/lib/rendering/svelteflow/layout/elkLayout.js]: Same behavior as before, but getLayoutedNodes() creates a new ELK instance with a new Web Worker on every call and never terminates it. Worth fixing now that the layout code is isolated in this file.

File: backend/src/test/java/org/rdfarchitect/services/update/CopyClassServiceTest.java [hunk 4]
```diff
@@ -72,6 +78,16 @@ void setUp() {
         databasePort.createGraph(graphIdentifier, graphSource.graph());
     }
 
+    private CopyClassResponseDTO copyClass(PasteClassesRequestDTO pasteRequest) {
+        var source = new PasteSourceClassDTO();
+        source.setSourceDatasetName(graphIdentifier.datasetName());
+        source.setSourceGraphURI(graphIdentifier.graphUri());
+        source.setClassUUID(CLASS_UUID);
+        pasteRequest.setSources(List.of(source));
```
Comment (spah-soptim) [inline backend/src/test/java/org/rdfarchitect/services/update/CopyClassServiceTest.java:86]: The CopyClassServiceTest helper always sets exactly one source (`List.of(source)`), the new multi-source `copyClasses()` path and the same-label-in-one-transaction de-duplication are never tested.
Maybe add regression test that pastes multiple sources, including a duplicate of the same class, in a single call.

File: frontend/src/routes/mainpage/packageNavigation/save-copy-class-to-backend.js [hunk 5]
```diff
@@ -36,27 +36,23 @@ export async function saveCopyClass(
     copyAttributes,
     copyAssociations,
 ) {
-    if (!copyState.classUUID || !copyState.graphURI || !copyState.datasetName)
-        return false;
+    const entries = copyState.getEntries();
+    if (entries.length === 0) return false;
```
Comment (spah-soptim) [inline frontend/src/routes/mainpage/packageNavigation/save-copy-class-to-backend.js:40]: saveCopyClass returns false for empty entries but undefined / does nothing on success and error.

File: frontend/src/lib/rendering/svelteflow/components/SvelteFlowPaneContextMenu.svelte
Comment (spah-soptim) [inline frontend/src/lib/rendering/svelteflow/components/SvelteFlowPaneContextMenu.svelte]: If packagesJSON.internalPackageList or packagesJSON.externalPackageList is null / absent, the paste fails silently.
Add same guard from Edit.svelte/getPackages (where it uses `?? []`) here for consistency.

File: frontend/src/routes/layout/menu-bar/Edit.svelte
Comment (spah-soptim) [inline frontend/src/routes/layout/menu-bar/Edit.svelte:190]: Pre-existing, but disablePasteButton includes `!editorState.selectedDiagram.getValue()`.

selectedDiagram is a StateObjectPair whose getValue() always returns the backing object (reset() just sets `{type:null, id:null}`), so this is always false and never disables paste. Likely intended to be `!editorState.selectedDiagram.getProperty("id")`.

File: frontend/src/lib/sharedState.svelte.js [hunk 6]
```diff
@@ -118,21 +199,39 @@ export const compareState = {
...
+
+    set(entries) {
+        this.entries.updateValue([...entries]);
+    },
+
+    removeByUuid(classUUID) {
+        this.entries.updateValue(
+            this.getEntries().filter(e => e.classUUID !== classUUID),
+        );
+    },
```
Comment (spah-soptim) [inline frontend/src/lib/sharedState.svelte.js]: copyState entries are identified by the triple `{classUUID, graphURI, datasetName}`, but `removeByUuid()` filters only on `classUUID`.
Not sure if it can happen, but if the clipboard ever holds the same class UUID from two different graphs, deleting it in one graph would also drop the other graph's clipboard entry.

Maybe add dataset and graph like the rest of the multi-select where it keys on dataset+graph+uuid.

File: frontend/src/routes/mainpage/packageNavigation/ClassEntry.svelte [hunk 7]
```diff
@@ -132,29 +332,31 @@
...
-            onclick={selectClass}
+            onclick={onEntryClick}
+            oncontextmenu={onEntryContextMenu}
         />
     </ContextMenu.TriggerArea>
     <ContextMenu.Content>
         {#if classType === ClassType.SINGLE_CLASS}
             <ContextMenu.Item.Button
                 onSelect={copyClass}
+                disabled={crossGraphDisabled}
```
Comment (spah-soptim) [inline frontend/src/routes/mainpage/packageNavigation/ClassEntry.svelte:344]: In the package-navigation context menu, Copy is disabled for cross-graph multi-selections, while paste fully supports cross-graph sources and the diagram context menu also allows cross-graph copy, which is inconsistent.

File: frontend/src/lib/rendering/svelteflow/interaction/nodeOrder.svelte.js [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+                zPosition: order.indexOf(id),
+            };
+        });
+
+        this.#bec.updateClassPositions(
+            editorState.selectedDataset.getValue(),
+            editorState.selectedGraph.getValue(),
+            editorState.selectedDiagram.getProperty("id"),
+            classPositionDTOList,
+        );
```
Comment (spah-soptim) [inline frontend/src/lib/rendering/svelteflow/interaction/nodeOrder.svelte.js]: Pre-existing, but NodeOrderController.#persistNodeOrder always calls bec.updateClassPositions with editorState.selectedGraph. For cross-graph (custom dataset) diagrams selectedGraph is null, so z-order changes made via the context menu persist to a 'null' graph URI and are effectively lost on reload. The drag-move path in svelteFlowWrapper.updateNodePositions handles this correctly by branching to updateGlobalClassPositions when selectedGraph is null.

File: backend/src/main/java/org/rdfarchitect/api/controller/datasets/diagrams/CustomDatasetDiagramAllClassesRESTController.java
Comment (spah-soptim) [inline backend/src/main/java/org/rdfarchitect/api/controller/datasets/diagrams/CustomDatasetDiagramAllClassesRESTController.java]: Pre-existing typo, but the logger says `DELETE` while this is `POST`.

**Review by spah-soptim — CHANGES_REQUESTED** [review]

File: frontend/src/routes/mainpage/packageNavigation/ClassEntry.svelte (hunk 7 above)
Comment (rema-soptim) [reply frontend/src/routes/mainpage/packageNavigation/ClassEntry.svelte:344]: This is intentional. I know we start transactions on multiple graphs to read, but when writing we would have the issue discussed with undo for multiple graphs at once. (RDFA-631)

**Review by spah-soptim — DISMISSED** [review]
LGTM, Review done. Smoke tested.

**Review by kiph-soptim — CHANGES_REQUESTED** [review]
- Selecting a class in the Merged View diagram, does not select the merged class in the navigation and the other way around
- It isn't possible to remove multiple classes from a custom dataset diagram if they are from different graphs. The backend only takes a list of class uuids, so it should be possible to remove them together
- It was possible to move the diagram with clicking the mouse wheel, is there a reason why it isn't possible anymore?

**Review by kiph-soptim — APPROVED** [review]
LGTM

**Review by spah-soptim — APPROVED** [review]
